pennylane_lightning/core/src/simulators/lightning_tensor/cutn/MPSCutn.hpp

```diff
@@ -0,0 +1,290 @@
+// Copyright 2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file MPSCutn.hpp
+ * MPS class with cutensornet backend.
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <complex>
+#include <vector>
+
+#include <cuda.h>
+#include <cutensornet.h>
+
+#include "CudaTensor.hpp"
+#include "CutnBase.hpp"
+#include "DataBuffer.hpp"
+#include "DevTag.hpp"
+#include "TensornetBase.hpp"
+#include "cuda_helpers.hpp"
+#include "cutnError.hpp"
+#include "cutn_helpers.hpp"
+
+/// @cond DEV
+namespace {
+namespace cuUtil = Pennylane::LightningGPU::Util;
+using namespace Pennylane::LightningGPU;
+using namespace Pennylane::LightningTensor::Cutn;
+using namespace Pennylane::LightningTensor::Cutn::Util;
+} // namespace
+/// @endcond
+
+namespace Pennylane::LightningTensor::Cutn {
+
+/**
+ * @brief Managed memory MPS class using cutensornet high-level APIs
+ * backed.
+ *
+ * @tparam Precision Floating-point precision type.
+ */
+
+template <class Precision>
+class MPSCutn final : public CutnBase<Precision, MPSCutn<Precision>> {
+  private:
+    using BaseType = CutnBase<Precision, MPSCutn>;
+
+    size_t maxBondDim_;
+
+    std::vector<std::vector<size_t>> sitesModes_;
+    std::vector<std::vector<size_t>> sitesExtents_;
+
+    std::vector<std::vector<int64_t>> sitesExtents_int64_;
+    std::vector<int64_t *> sitesExtentsPtr_int64_;
+    std::vector<void *> tensorsDataPtr_;
```

Comment:
Not the biggest fan of these vectors of pointers. Who owns the data we expect to exist in here? Where is this data allocated/freed?

Is there another way we can store the information here? Especially with vector<void *>. If we want to store pointer addresses, I'd prefer `std::vector<uint64_t>`, or another alternative instead. What happens if the pointers are deleted/moved/etc?

Reply:
Good point! In the updated implementation, instead of storing `ptrs` to `tensors_`'s `cudaTensor` and `sitesExtents_int64_`'s data into a private variables, we rely on `getSitesExtentsPtr() ` to return the `ptr` to  `sitesExtents_int64_` and `getTensorsDataPtr()` to return the `ptr` to `tensor_`. This implementation will ensure that the pointers points to the right address even when changes are made to `tensor_`.  Replace std::vector<void*> with std::vector<uint64_t*>